src/creeps/Zombie.cpp

```diff
@@ -1,61 +1,308 @@
+            break;
+    }
+
+    if (frame > 0) {
+        --frame;
+    } else {
+        setCurFrame(ZOMBIE_FRAMES);
+        ++step;
+    }
+
+    setCurDir(d);
+    if (state != ZombieState::ZOMBIE_MOVE) {
+        setState(ZombieState::ZOMBIE_MOVE);
+    }
 }
 
-void Zombie::generateRandomMove(){
-    double cosVal = cos(getAngle()*PI/180.0);
-    double sinVal = sin(getAngle()*PI/180.0);
+/**
+ * A* algo generates a string of direction digits.
+ * Fred Yang
+ * Feb 14
+ */
+string Zombie::generatePath(const float xStart, const float yStart,
+        const float xDest, const float yDest) {
+    // priority queue index
+    int index = 0;
+
+    // temp index
+    int i, j;
+
+    // row / column index
+    int x, y, xdx, ydy;
 
-    const int x = getVelocity();
-    const int y = getVelocity();
+    // temp char
+    char c;
 
-    const int tx = getX() + 20 * cosVal;
-    const int ty = getY() + 20 * sinVal;
+    // path to be generated
+    string path;
 
-    movementHitBox->move(tx,ty);
-    projectileHitBox->move(tx,ty);
+    // current node & child node
+    static Node curNode;
+    static Node childNode;
 
-    CollisionHandler &ch = GameManager::instance()->getCollisionHandler();
-    if(ch.detectMovementCollision(this)){
-        setAngle(getRandomAngle());
-        cosVal = cos(getAngle() * PI / 180.0);
-        sinVal = sin(getAngle() * PI / 180.0);
+    // priority queue
+    static priority_queue<Node> pq[2];
+
+    // reset the node maps
+    memset(closedNodes, 0, sizeof(closedNodes[0][0]) * ROW * COL);
+    memset(openNodes, 0, sizeof(openNodes[0][0]) * ROW * COL);
+
+    const int xNodeStart = static_cast<int>(xStart / TILE_SIZE);
+    const int yNodeStart = static_cast<int>(yStart / TILE_SIZE);
+    const int xNodeDest = static_cast<int>(xDest / TILE_SIZE);
+    const int yNodeDest = static_cast<int>(yDest / TILE_SIZE);
+
+    // create the start node and push into open list
+    curNode = Node(xNodeStart, yNodeStart, 0, 0);
+    curNode.updatePriority(xNodeDest, yNodeDest);
+    pq[index].push(curNode);
+
+    // A* path finding
+    while (!pq[index].empty()) {
+        Node tmp = pq[index].top();
+        // get the current node with the highest priority from open list
+        curNode = Node(tmp.getXPos(), tmp.getYPos(), tmp.getLevel(), tmp.getPriority());
+
+        x = curNode.getXPos();
+        y = curNode.getYPos();
+
+        // remove the node from the open list
+        pq[index].pop();
+        openNodes[x][y] = 0;
+
+        // mark it on the closed nodes map
+        closedNodes[x][y] = 1;
+
+        // quit searching when the destination is reached
+        if (x == xNodeDest && y == yNodeDest) {
+            // generate the path from destination to start
+            // by following the directions
+            path = "";
+            while (!(x == xNodeStart && y == yNodeStart)) {
+                j = dirMap[x][y];
+                c = '0' + (j + DIR_CAP / 2) % DIR_CAP;
+                path = c + path;
```

Comment:
'c' is only ever used and assigned in this loop
It's not necessary, so remove c and replace with its value in this line